examples/minimal/packages/client-react-external-wallet/src/mud/customWalletClient.ts

```diff
@@ -0,0 +1,82 @@
+import type { WriteContractParameters, Chain, Account, WalletActions } from "viem";
+import { sendTransaction, writeContract } from "viem/actions";
+import { useAccount, useWalletClient, type UseWalletClientReturnType } from "wagmi";
+import pRetry from "p-retry";
+import { getNonceManager } from "@latticexyz/common";
+import { useMUDNetwork, type MUDNetwork } from "./NetworkContext";
+
+export const useMUD = () => {
+  const network = useMUDNetwork();
+  const { data: connectorWalletClient } = useWalletClient();
+  const { chainId } = useAccount();
+
+  let walletClient;
+  if (network.publicClient.chain.id === chainId && connectorWalletClient?.chain.id === chainId) {
+    // TODO: Should this be memoized?
+    // `walletClient = connectorWalletClient.extend(burnerActions);` is unnecessary for an external wallet
+    walletClient = connectorWalletClient.extend(setupObserverActions(network.onWrite));
+  }
+
+  return { network, walletClient };
+};
+
+export type WalletClient = NonNullable<UseWalletClientReturnType["data"]>;
+
+// See @latticexyz/common/src/sendTransaction.ts
+const burnerActions = (client: WalletClient): Pick<WalletActions<Chain, Account>, "sendTransaction"> => {
+  // TODO: Use the `debug` library once this function has been moved to the `common` library.
+  const debug = console.log;
+
+  return {
+    sendTransaction: async (args) => {
+      const nonceManager = await getNonceManager({
+        client,
+        address: client.account.address,
+        blockTag: "pending",
+      });
+
+      return nonceManager.mempoolQueue.add(
+        () =>
+          pRetry(
+            async () => {
+              if (!nonceManager.hasNonce()) {
+                await nonceManager.resetNonce();
+              }
+
+              const nonce = nonceManager.nextNonce();
+              debug("sending tx with nonce", nonce, "to", args.to);
+              return sendTransaction(client, { ...args, nonce } as typeof args);
```

Comment:
we should prob use `getAction('sendTransaction')` here instead of importing from viem, in case the client has overridden this method already

Reply:
> session wallet

Yes, we should use all custom actions (nonce, retry, write$) for session wallets.

In the upcoming burner wallet PR #2309, all these actions are applied to the burner wallet. When using both external and burner wallets, I believe the external wallet doesn't need to be extended. This is because, we don't require nonce management for MetaMask, and dev-tools, which takes a single wallet client, should be used with the burner wallet.

---

> getAction

Right, using getAction is better since it accounts for more extensions beforehand. I'll switch to getAction.